### PR TITLE
[feature] Perform backups (and restores!) into EPFL's Scality® object storage system

### PR DESCRIPTION
**Pull request by domq**
We use [restic](https://restic.net/) to back up and restore user-important data, i.e. the contents of the database, the `wp-content` directory, and the `.htaccess` file. All the steps, including the [necessary rewriting and invalidation operations](https://confluence.epfl.ch:8443/pages/viewpage.action?pageId=125734816), are taken care of by Ansible.

Scality is a vendor of an object storage system, that provides an Amazon® S3-compatible API. (All trademarks are properties of their respective owners, yadda yadda)

Restoring across namespaces is possible in a single direction (from production to tests); to this effect, read-only access is granted to members of `/keybase/team/epfl_wp_test` by sharing the appropriate credentials on a need-to-know basis.

A complete backup and restore cycle across namespaces (from production to tests) can be demoed with the following commands:

```
./ansible/wpsible --prod -t wp.backup -l www__labs__dcsl,test_migration_wp__labs__dcsl -vv

./ansible/wpsible -t wp.wipe,wp.restore,wp -l test_migration_wp__labs__dcsl -vv \
  -e '{"wp_destructive" : { "test_migration_wp__labs__dcsl": ["wipe"] }, "wp_restore_from_namespace": "wwp", "wp_restore_from_name": "www__labs__dcsl"}'
```

Although this is yet to be tested, provisions have been made so that this code will work on AWX, where scheduled backups will be eventually implemented.

Summary of changes:

- Normalize all tags in the `wordpress-instance` role: bring them under the `wp.*` namespace, keeping the obsolete ones for a while; fix a number of bugs related to implicit tags (i.e. use a consistent doctrine of either `import_tasks` + `tags:` or `include_tasks` + `tags:` + `apply:` throughout)
- Rewrite `backup.yml` and `restore.yml` in terms of restic; abandon `restore-over-ssh.yml`
- Propagate backup credentials using `lookup("env_secrets", ...)` introduced in #321, paving the way for usage on AWX
- Pass said credentials through to the interactive shells created by ssh'ing into mgmt, so that people can wield `restic` command lines without further ado e.g. (on `wwp-test`) <pre>restic -r s3:https://s3.epfl.ch/svc0041-47680839aba24f7c02f160cd9d58a5e1/backup/wordpresses/test_migration_wp__labs__dcsl/files snapshots</pre>
- Refactor the `lookup("env_secrets", ...)` mechanism a bit, so that it now allows “separate compilation”. Previously, in order for any env-secret lookup to work, all variables for all env-secrets needed to be provided at once (because Ansible doesn't believe in lazy evaluation of dicts). The `env_secret_*` variables are now separate from each other (instead of ~being fields of the same `env_secrets` dict~), allowing each to be evaluated even if the dependencies of the others are not loaded (through `include_vars`)

Details on the restore implementation:

- Back up files and SQL dumps of each WordPress instance into separate S3 folders. Although restic does support multiplexing backup streams into a single bucket, its tagging scheme feels “20th-century like” i.e. it is limited to `--host` and `--path`,  which is not adequate for our use case¹ 
- Refactor the initialization sequence for stateful and “connect-ful” plug-ins (epfl-menus, mainwp-child ewww-image-optimizer), whose state may change during relocation, so that said initialization sequence occurs both during site creation and site restore. During a restore, leave the other plug-ins and their state alone (i.e. technically, restore the symlinks in `wp-content/themes` and `wp-content/plugins` as part of the files restore); thus minimizing downtime, not to mention cognitive overload when reasoning about the restore sequence
- Restore .htaccess and make sure WordPress rewrites its section thereof (`wp  rewrite flush --hard`); the rest of that file being out of scope for Ansible automation (at least for now)
- Be sure **not** to restore `wp-config.php`.  When restoring after a wipe (as in the example command line above), a completely new database and directory will be populated with the old data, thereby achieving “software rejuvenation” and avoiding pitfalls such as aliasing the restored site's database to the original site's (ouch!)²

¹ Specifically, we are unable to predict `--path` from `wp_restore_from_name`. According to [this forum post](https://forum.restic.net/t/what-are-parent-snapshots/26/6) we could abuse `--host` without too much of a penalty (i.e. useless reads but *not* lost space in useless reverse-deltas), but this seems quite inelegant...
² This happened during the development of this PR, but fortunately we had a backup — In S3, natch... We can therefore proudly proclaim that the restic backups were already used once in a real-life disaster recovery situation, even before this PR was pushed in full to GitHub, let alone merged!

### CODE REVIEW

**Comment by domq** [comment]
@jdelasoie contrary to what I mentioned in our discussion today, the `wp.backup` actions always were an implicit dependency of `wp.wipe`; but such was not the case between `wp.restore` and `wp.create`, because I had made a conscious effort to change that. I reverted that part, now `wp.wipe` implies `wp.backup` and `wp.restore` implies `wp.create` once again. PTAL

**Comment by domq** [comment]
I just rebased this on top of #325; no other changes.

File: ansible/roles/wordpress-instance/tasks/restore.yml [hunk 1]
```diff
@@ -1,15 +1,100 @@
...
+
+- name: Restore database
+  environment: "{{ restore_restic_environment }}"
+  shell:
+    cmd: |
+      restic -r {{ restore_restic_repo_sql }} --no-lock --quiet dump latest db-backup.sql | \
+        {{ wp_cli_command }} db import -
+  tags: wp.restore.sql
+
+#### Cleanup tasks ####
```
Comment (jdelasoie) [inline ansible/roles/wordpress-instance/tasks/restore.yml:62]: Je ne perçois pas ces actions comme du cleanup, mais plutôt du 
```suggestion
#### Post-restore tasks ####
```

**Review by jdelasoie — APPROVED** [review]

File: ansible/roles/wordpress-instance/tasks/plugins.yml [hunk 2]
```diff
@@ -212,15 +212,19 @@
     {{ wp_cli_command }} pll lang list --field=term_id --format=csv
   changed_when: false
   register: _polylang_languages_csv
-  tags: plugins.polylang
+  tags:
+    - wp.plugins.polylang
+    - plugins.polylang
```
[the remark this answers is not in the document]
Comment (domq) [reply ansible/roles/wordpress-instance/tasks/plugins.yml:217]: En effet, mais je pense qu'il faut attendre une prochaine PR pour nettoyer les vieux noms; ils sont encore en usage par ex. dans AWX

File: ansible/roles/wordpress-instance/tasks/backup.yml [hunk 3]
```diff
@@ -1,56 +1,49 @@
...
-         --exclude "ansible-backup-*" -clf - .) | \
-      (cd "{{ _backup_tmpdir }}"; tar xpvf -)
-    cp "{{ wp_dir }}/.htaccess" "{{ _backup_tmpdir }}/htaccess" || true
-  tags:
-    - wipe
-    - backup
-    - backup.code
+# Backup a WordPress instance to EPFL's Scality system through its S3-like API
+
+- assert:
```
[the remark this answers is not in the document]
Comment (domq) [reply ansible/roles/wordpress-instance/tasks/backup.yml:3]: Après RTFM, je reviens sur ma déclaration concernant AWX et les CallbackModules; cependant je préfèrerais aborder ce problème dans une autre PR.

**Comment by domq** [comment]
All review comments heeded, PTAL @jdelasoie

**Review by jdelasoie — APPROVED** [review]

**Review by ponsfrilus — COMMENTED** [review]
Thanks for the tags renaming !

**Pull request merged**